bittide-instances/src/Bittide/Instances/Hitl/SwCcTopologies.hs

```diff
@@ -570,154 +477,217 @@ topologyTest refClk sysClk sysRst IlaControl{syncRst = rst, ..} rxNs rxPs miso c
-  success = not <$> (transceiversFailedAfterUp .||. startBeforeAllReady)
+  ugns :: Vec LinkCount (Signal GthTx (BitVector 64))
+  ugns = zipWith (-) txCounters rxCntrs
+  -- see NOTE [magic start values]
 
-  skip =
-    register
-      sysClk
-      sysRst
-      enableGen
-      False
-      (maybe False (not . fpgaEnabled) <$> testConfig)
+  ugns' :: Vec LinkCount (Signal Basic125 (BitVector 64))
+  ugns' = zipWith3 go transceivers.txClocks othersNotInCCResetSync ugns
+   where
+    go txClk enaSig =
+      regEn
+        sysClk
+        clockControlReset
+        enableGen
+        rxCounterStartUgn
+        enaSig
+        . xpmCdcArraySingle txClk sysClk
+
+  ugnsStable :: Vec LinkCount (Signal Basic125 Bool)
+  ugnsStable = go <$> ugns'
+   where
+    go ugn = ugnStable
+     where
+      ugn' = bitCoerce <$> ugn
+      stabInd =
+        withClockResetEnable
+          sysClk
+          syncRst
+          enableGen
+          $ SI.stabilityChecker
+            (SNat @CccStabilityCheckerMargin)
+            (SNat @(CccStabilityCheckerFramesize GthTx))
+            ugn'
+      ugnStable = SI.stable <$> stabInd
+
+  maskWithCfg ::
+    Bool ->
+    Vec LinkCount (Signal Basic125 Bool) ->
+    Signal Basic125 (Vec LinkCount Bool)
+  maskWithCfg dflt = liftA2 go1 (mask <$> cfg) . bundle
+   where
+    go1 m = zipWith go2 (bitCoerce m)
+    go2 m val = if m then val else dflt
 
-  testConfig :: Signal Basic125 (Maybe TestConfig)
-  testConfig = hitlVio disabled sysClk done success
+  allUgnsStable =
+    trueFor (SNat @(Seconds 2)) sysClk clockControlReset
+      $ and
+      <$> maskWithCfg True ugnsStable
 
-makeTopEntity 'hwCcTopologyWithRiscvTest
+  findFifoError bits = result
+   where
+    masked = maskWithCfg False bits
+    observingError = or <$> masked
+    observedError = sticky sysClk clockControlReset observingError
+    result = not <$> observedError
+
+  noFifoOverflows = findFifoError fifoOverflowsFree
+  noFifoUnderflows = findFifoError fifoUnderflowsFree
+
+  -- NOTE [magic start values]
+  -- These values could be anything, but are chosen to be recognisable and help debugging.
+  --   0x2abbccddeeff1234 - 0x1122334411223344 = 0x19999999dddcdef0
+  -- If you ever see the ugn being a constant 0x19999999dddcdef0
+  -- then you know the your counter isn't running and you're receiving 'Nothing',
+  -- If you see 0x19999999.......... and it's counting up, then you're receiving Nothing,
+  -- but your counter is running.
+
+  ( ugn0
+    , ugn1
+    , ugn2
+    , ugn3
+    , ugn4
+    , ugn5
+    , ugn6
+    ) = vecToTuple ugns'
+
+  ( dDiff0
+    , dDiff1
+    , dDiff2
+    , dDiff3
+    , dDiff4
+    , dDiff5
+    , dDiff6
+    ) = vecToTuple domainDiffs
+
+type WaitCycles dom hold prd = PeriodToCycles dom (prd - hold)
+
+data ToFincFdecState' dom hold min
+  = Wait (Index (WaitCycles dom hold min))
+  | Pulse (Index (PeriodToCycles dom hold)) SpeedChange
+  | Idle
+  deriving (Generic, NFDataX, Eq)
+
+speedChangeToFincFdec' ::
```

Comment:
Same here. Typically `'` is used for "strict" versions of a function, but that's not the case here.